common/OpTestSSH.py

```diff
@@ -155,52 +159,133 @@ def get_console(self):
-        except pexpect.TIMEOUT as e:
-            print e
-            print "# TIMEOUT waiting for command to finish."
-            print "# Attempting to control-c"
-            try:
-                console.sendcontrol('c')
-                rc = console.expect(["\[console-pexpect\]#$"], 10)
-                if rc == 0:
-                    raise CommandFailed(command, "TIMEOUT", -1)
-            except pexpect.TIMEOUT:
-                print "# Timeout trying to kill timed-out command."
-                print "# Failing current command and attempting to continue"
-                self.terminate()
-                raise CommandFailed("ssh -p %u" % self.port, "timeout", -1)
-            raise e
+    def retry_password(self, console, command):
+        retry_list_output = []
+        a = 0
+        while a < 3:
+          a += 1
+          console.sendline(self.password)
+          rc = console.expect([".*#", "try again.", pexpect.TIMEOUT, pexpect.EOF])
+          if (rc == 0) or (rc == 1):
+            combo_io = console.before + console.after
+            retry_list_output += combo_io.splitlines()
+            matching = [xs for xs in sudo_responses if any(xs in xa for xa in console.after.splitlines())]
+            if len(matching):
+              echo_rc = 1
+              rc = -1 # use to flag the failure next
+          if rc == 0:
+            retry_list_output += self.set_env(console)
+            echo_rc = 0
+            break
+          elif a == 2:
+            echo_rc = 1
+            break
+          elif (rc == 2) or (rc == 3):
+            raise CommandFailed(command, 'Retry Password TIMEOUT ' + ''.join(retry_list_output), -1)
+        return retry_list_output, echo_rc
 
+    def handle_password(self, console, command):
+        # this is for run_command 'sudo -s' or the like
+        handle_list_output = []
+        failure_list_output = []
+        pre_combo_io = console.before + console.after
+        console.sendline(self.password)
+        rc = console.expect([".*#", "try again.", pexpect.TIMEOUT, pexpect.EOF])
+        if (rc == 0) or (rc == 1):
+          combo_io = pre_combo_io + console.before + console.after
+          handle_list_output += combo_io.splitlines()
+          matching = [xs for xs in sudo_responses if any(xs in xa for xa in console.after.splitlines())]
+          if len(matching):
+            # remove the expect prompt since matched generic #
+            del handle_list_output[-1]
+            echo_rc = 1
+            rc = -1 # use this to flag the failure next
+        if rc == 0:
+          # with unknown prompts and unknown environment unable to capture echo $?
+          echo_rc = 0
+          self.set_env(console)
+          list_output = handle_list_output
+        elif rc == 1:
+          retry_list_output, echo_rc = self.retry_password(console, command)
+          list_output = (handle_list_output + retry_list_output)
+        else:
+          if (rc == 2) or (rc == 3):
```

Comment:
Here also handle EOF separately, catch that and set the state to disconnected. After that raise "Session EOF" exception.

Reply:
Fixed

Reply:
Thanks